### DISCUSSION TITLE
Select Value shows undefined

### DISCUSSION THREAD

**Discussion by KJ-tasktop** [react-hook-form/react-hook-form#5291] [Q&A] [answered]
I am hopeful that I can get some help with this. I feel like I can see the light at the end of a very dark tunnel, but this part is taking hours, and I submit... it's tough to make a CSB because this is pretty complicated, so I'll do my best to explain it. Our Marketing department uses Marketo forms which do not play well with our stack. RHF seems like an excellent solution, but all of the examples that I have found seem based on the idea that you're actually coding a form that you know in advance what the fields will be. I need something more dynamic, so I wrote a script that will scrape the Marketo form, and present the fields through RHF. All the Marketing team needs to do is enter a form ID into the CMS. In the front end, the component that handles this is called MktoForm. It calls a hook that injects the Marketo form, which it scrapes for an array of all of the visible fields. The array is sent to the RHF component which generates a form from those fields: 

`<form onSubmit={handleSubmit(submitIt)}>
    {customForm.map((r, i) => (
        <React.Fragment key={`${r.node}-${i}`}>
            {r.node === 'input' && (
                <>
                    <Input key={r.id} register={register} errors={errors} {...r} />
                </>
            )}

            {r.node === 'textarea' && (
                <TextArea key={r.id} register={register} errors={errors} {...r} />
            )}
            {r.node === 'select' && (
                <Select
                    key={r.id}
                    register={register}
                    errors={errors}
                    control={control}
                    handleSelect={(e) => updatedSelect(e)}
                    {...r}
                />
            )}
        </React.Fragment>
    ))}
    {checkBox ? <OptIn register={register} errors={errors} /> : null}
    <button type="submit">Submit</button>
</form>`

These components all live in a separate file called FormComponents, and so far, so good. The issue is with conditional rendering on the Select component. I didn't want to use react-select, but it seems like it became a necessity when I wanted an onChange. 

`const Select = ({ errors, handleSelect, control, ...r }) => {
	const { copy, required, options } = r
	const validation = {
		required: required ? `${copy} is required` : false,
	}

	return (
		<FormGroup className="select">
			<Controller
				id={copy}
				name={copy}
				isClearable
				isSearchable
				control={control}
				render={({ field }) => (
					<ReactSelect {...field} options={options} onChange={(e) => handleSelect(e)} theme={customTheme} />
				)}
			/>
			{errors[copy] && errors[copy].type === 'required' && <Error>{errors[copy].message}</Error>}
			<Label>{copy}</Label>
		</FormGroup>
	)
}`

There are several use cases, but the one that is on the test form is if the user selects United States as a country, an additional field should appear that prompts them to select a state. I did find that this was working correctly, but only if the form was submitted - the Marketo form (which is display none) would throw an error, and RHF would re-render, but would include the additional field which was now visible on the Marketo form. So, to handle this, I set up an onChange on the Select that would get the form values, and cause the MktoForm to re-render. I am able to print out the values for both forms. 

Here's what I get from RHF: 
<img width="654" alt="Screen Shot 2021-05-22 at 11 31 32 AM" src="https://user-images.githubusercontent.com/75387705/119237364-473d0c80-baf1-11eb-8f38-14185e12604f.png">

There is another script that I am using that is adapted to my stack that was originally written by Sandford Whiteman who is on all of the Marketo boards. It does pick up the value from the RHF Select, and passes it to the Marketo form - but I can't actually submit the RHF form to trigger the submit of the Marketo form because it fails on the validation. Here is the data that is picked up by the Marketo script: 

<img width="628" alt="Screen Shot 2021-05-22 at 11 39 37 AM" src="https://user-images.githubusercontent.com/75387705/119237577-67210000-baf2-11eb-8a62-c0e7f179cbc0.png">

It seems like it's going to be a really simple fix, but I just can't seem to figure it out. I've gone through the docs, about a million stack overflow questions, and through the discussions here. I could really use a hand!

**Comment by bluebill1049** [comment, marked as answer]
can you provide a codesandbox and provide steps to reproduce?

**Reply by KJ-tasktop** [reply]
Well, not exactly, as it's pretty complicated. I did set up a CSB, but it seems to be working correctly - although, maybe the issue is that it's returning a ReactSelect object. I can't be sure. It is here: https://codesandbox.io/s/aged-haze-i4hfg?file=/src/index.js:565-599

 In the CSB, I have added my Select, and the function to print the data on change. I have also included my actual code, but it's not being called anywhere because it won't work because I can't find a way to embed a Marketo form without exposing some variables that the boss man would undoubtedly prefer that I don't expose. 

Since my original post, I have tried the setValue function which seemed really promising, but I don't know whether to laugh or cry... now it shows the selection in the RHF console data, but it doesn't catch it in the Marketo scrape. 

<img width="644" alt="Screen Shot 2021-05-23 at 1 55 34 AM" src="https://user-images.githubusercontent.com/75387705/119254226-fa494c80-bb69-11eb-8fd8-2eef7e2cd03c.png">

<img width="661" alt="Screen Shot 2021-05-23 at 1 56 04 AM" src="https://user-images.githubusercontent.com/75387705/119254234-0c2aef80-bb6a-11eb-8209-235a8c250e76.png">

I have noticed as well that while I'm using setValues, the React Select doesn't actually show the value, it just stays on the placeholder. This is also the case in the CSB. The data is null here, but that's because setData is never called. I'm not sure how it works with getValues. 

The Select component is in FormComponents.js starting at line 116, and it's called from ReactForm.js at line 72. The form itself comes from line 123 of MktoForm.js. 

Thanks a lot for helping with this, Bill. I am usually pretty hesitant to reach out in this way, but I have been stuck with this problem for a couple of days, and I'm really losing my mind over it. RHF has helped immensely with building forms, and I do greatly appreciate the work that you've put into it.

**Reply by bluebill1049** [reply]
hey, I curious why not using `fields.onChange` instead of manually setting the value? 

https://codesandbox.io/s/boring-darkness-en1zi?file=/src/index.js

<img width="728" alt="Screen Shot 2021-05-23 at 7 11 54 pm" src="https://user-images.githubusercontent.com/10513364/119254633-bfdbc000-bbfa-11eb-8f2e-bd6f7add6423.png">

**Reply by KJ-tasktop** [reply]
Cool! That gave me back an object, but it has the value that I needed, and I am able to refactor the rest accordingly. Thanks, Bill! I really do appreciate it!

**Reply by bluebill1049** [reply]
fanatsic.
